src/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -291,21 +294,23 @@ internal PipeAwaiter<FlushResult> FlushAsync(CancellationToken cancellationToken
                     CommitUnsynchronized();
                 }
 
-                awaitable = _readerAwaitable.Complete();
+                _readerAwaitable.Complete(out completion, out completionState, out pipeScheduler);
 
                 cancellationTokenRegistration = _writerAwaitable.AttachToken(cancellationToken, s_signalWriterAwaitable, this);
             }
 
             cancellationTokenRegistration.Dispose();
 
-            TrySchedule(_readerScheduler, awaitable);
+            TrySchedule(pipeScheduler ?? _readerScheduler, completion, completionState);
 
-            return new PipeAwaiter<FlushResult>(_writer);
+            return new ValueTask<FlushResult>(_writer);
```

Comment:
At some point I'd recommend looking at whether you can have a fast-path that uses `new ValueTask<FlushResult>(result)` rather than always wrapping the `IValueTaskSource`.  There can be a very meaningful performance difference on hot paths between the two.

Reply:
Agreed, there are likely a few places where we can do that.